### PR TITLE
Improve ergonomic of the Pytorch dataset and Generate embeddings for oxford pet

### PR DESCRIPTION
**Pull request by eddyxu**
* Improve ergnomic of Pytorch Dataset to load local directory
* Auto tune learning rate and uses other hyperparameters from torchvision site.
* Example code to generate embeddings.

### CODE REVIEW

File: python/benchmarks/oxford_pet/embeddings.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        for batch, pk in train_loader:
+            batch = batch.to("cuda")
+            features = extractor(batch)["features"].squeeze()
+            df = pd.DataFrame(
+                {
+                    "pk": pk,
+                    "features": features.tolist(),
+                }
+            )
+            dfs.append(df)
```
Comment (changhiskhan) [inline python/benchmarks/oxford_pet/embeddings.py:86]: does fast.ai / huggingface have any conveniences for this?

File: python/benchmarks/oxford_pet/train.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+import torchvision.transforms as T
+from common import Classification, RawOxfordPetDataset, collate_fn, raw_collate_fn
+from pytorch_lightning.loggers import TensorBoardLogger
+from torchdata.datapipes.iter import IterableWrapper
+from torchvision.transforms.functional import InterpolationMode
+
+import lance
+import lance.pytorch.data
+
+NUM_CLASSES = 38
```
Comment (changhiskhan) [inline python/benchmarks/oxford_pet/train.py:23]: duplicate?

File: python/benchmarks/oxford_pet/train.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+
+class TrainTransform(torch.nn.Module):
+    """
+    Image transform for training.
+
+    Adding random argumentations.
```
Comment (changhiskhan) [inline python/benchmarks/oxford_pet/train.py:30]: augmentation

File: python/benchmarks/oxford_pet/train.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+    https://github.com/pytorch/vision/blob/a89b1957a62e2f68f001d5d60268743edbe164d8/references/classification/presets.py#L6
+    """
+
+    def __init__(
+        self,
+        crop_size: float,
+        mean: tuple[float] = (0.485, 0.456, 0.406),
```
Comment (changhiskhan) [inline python/benchmarks/oxford_pet/train.py:38]: where do these defaults come from?

File: python/benchmarks/oxford_pet/embeddings.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        dataset = lance.pytorch.data.LanceDataset(
+            dataset,
+            columns=["image", "filename"],
+            batch_size=batch_size,
+            # filter=(pc.field("split") == "train")
+        )
+        train_loader = torch.utils.data.DataLoader(
+            dataset, num_workers=num_workers, batch_size=None, collate_fn=collate_fn
+        )
+    elif data_format == "raw":
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply python/benchmarks/oxford_pet/embeddings.py:60]: will add support for parquet later. Maybe tfrecord as well.

File: python/benchmarks/oxford_pet/common.py [hunk 6]
```diff
@@ -1,38 +1,18 @@
...
 import torch
 import torchvision
-import torchvision.transforms as T
-from PIL import Image
-from torch import optim
-from torchvision.models.efficientnet import EfficientNet_B0_Weights
-
-import lance
 import lance.pytorch.data
+from PIL import Image
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply python/benchmarks/oxford_pet/common.py:13]: Ok, lemme change it to `PIL.Image`

File: python/benchmarks/oxford_pet/common.py [hunk 7]
```diff
@@ -1,38 +1,18 @@
...
-"""Train and evaluate models on Oxford pet dataset.
-
-"""
+#!/usr/bin/env python3
 
 import io
 import os
 import time
-from typing import Callable, Optional
+from typing import Optional, Callable
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply python/benchmarks/oxford_pet/common.py:6]: Fixed

File: python/benchmarks/oxford_pet/common.py [hunk 8]
```diff
@@ -1,38 +1,18 @@
...
-
-
-def raw_collate_fn(batch):
-    images = []
-    labels = []
-    for img, label in batch:
-        images.append(img)
-        labels.append(label)
-    return torch.stack(images), torch.tensor(labels)
+NUM_CLASSES = 38
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply python/benchmarks/oxford_pet/common.py:15]: This is specific number on the dataset. I feel that it is overkill to calculate it dynamically via dataset. We especially need to support different formats (i.e., it requires some effort to dynamically compute this number in the raw format.).

File: python/benchmarks/oxford_pet/datagen.py [hunk 9]
```diff
@@ -15,12 +15,14 @@
 
 import os
 import pathlib
+import sys
+
+sys.path.append("..")
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply python/benchmarks/oxford_pet/datagen.py:20]: This is for import `bench_utils`.

File: python/benchmarks/oxford_pet/embeddings.py (hunk 1 above)
Comment (eddyxu) [reply python/benchmarks/oxford_pet/embeddings.py:86]: Since the rest is using pytorch lightning, using fast.ai / hf seems need to convert them to raw pytorch and make it adapt fastai/hf?

Prob we can use https://pytorch-lightning.readthedocs.io/en/stable/deploy/production_basic.html 

It still needs to match pk tho. 

I can make it to use pytorch lightning's predict if desired.

File: python/benchmarks/oxford_pet/train.py (hunk 2 above)
Comment (eddyxu) [reply python/benchmarks/oxford_pet/train.py:23]: Fixed.

File: python/benchmarks/oxford_pet/train.py (hunk 3 above)
Comment (eddyxu) [reply python/benchmarks/oxford_pet/train.py:30]: Done

File: python/benchmarks/oxford_pet/train.py (hunk 4 above)
Comment (eddyxu) [reply python/benchmarks/oxford_pet/train.py:38]: These numbers are used across torchvision, and used in the python code referred above.

File: python/benchmarks/bench_utils.py [hunk 10]
```diff
@@ -39,7 +40,11 @@ def read_file(uri) -> bytes:
     return fs.open_input_file(key).read()
 
 
-def download_uris(uris: Iterable[str], func=read_file) -> Iterable[bytes]:
+def download_image(uri: str) -> Image:
+    return ImageUri(uri).to_embedded()
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply python/benchmarks/bench_utils.py:44]: Ok, i will try it now.

File: python/benchmarks/oxford_pet/embeddings.py (hunk 1 above)
Comment (eddyxu) [reply python/benchmarks/oxford_pet/embeddings.py:86]: Oh, actually, this need to wrap into a separate module, as it uses `create_feature_extractor(model.backbone, {"avgpool": "features"})` feature extractor, while the original module should do basic predictions (i.e, just return detected class).

File: python/benchmarks/oxford_pet/common.py [hunk 11]
```diff
@@ -127,73 +108,37 @@ def training_step(self, batch, batch_idx):
...
-            dataset,
-            num_workers=num_workers,
-            batch_size=batch_size,
-            collate_fn=raw_collate_fn,
-        )
-    else:
-        raise ValueError("Unsupported data format")
+def collate_fn(transform):
+    def _collate_fn(batch):
+        # TODO: Labels should be converted via torch.LanceDataset
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply python/benchmarks/oxford_pet/common.py:127]: I have not try to create label type with the fixed mapping between label -> text yet.

Lemme try.

File: python/benchmarks/bench_utils.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (eddyxu) [reply python/benchmarks/bench_utils.py]: As discussed in #157 , use downloaded binary for now.

**Review by changhiskhan — APPROVED** [review]
Lgtm
